### DISCUSSION TITLE
RSA Encryption

### DISCUSSION THREAD

**Discussion by lintangtimur** [lintangtimur/ovoid#54] [Ideas]
### Flow enkripsi V3 accounts/login

Entah ini benar apa engga, flownya:
1. Data yang di encrypt ada deviceId, kyc, notelp
2. Data dijadikan json
3. json di encrypt dengan RSA public key, didapat base64
4. Lalu di sign dengan private key, output base64

![image](https://user-images.githubusercontent.com/16686825/121908459-3a5db280-cd57-11eb-8fa7-6224a6650caf.png)

![image](https://user-images.githubusercontent.com/16686825/121909707-63cb0e00-cd58-11eb-83d7-2d3760099cb9.png)

Bisa dikoreksi jika ada yang salah

**Comment by ii64** [comment]
Tambahan untuk data nya `signFirstTime`



```
datSharing = DataSharing(deviceId, kyc, phoneNumber, signFirstTime)

json = JSON.stringify(datSharing)
var sjson string

// encrypt data with RSA/ECB PKCS1-Pad - x509 pub
// result base64 encoded
var x509key = "MIIBIjANBgkqhkiG9w0BAQEFAAOCAQ8AMIIBCgKCAQEAgRjK/3xgE0NsWJ5AK+RTAqWVghZH83Fhnxu2lSvcDUTXUhCyhlKGgDIWWswe/S7cWiuPYebHoo0vp2bHUQLHNFfyzh/IWAaR+uF1gunu2YtH05CB7Jbb+ePajiBIsmxtmXHMuR1oiL2vISyrX1aV2xdewL8Tac9LhNroMMuojp+ezh+gIzePA/OSiyrc0WZwljQmLIAG/XtOU7HKHaAGEkZ5d231l5WwDwRbeGOTEn4mN6cwRtwNk5Q1kYxaI51Q1WBC9oN21jY9+3PSJctYbAIjhzHa3A+n79+2hQeFiXtElwxmwDcrC3mtWZn2L22lvoHln44PSftP4J3tJJP0pwIDAQAB"
var pubKey = KeyFactory.getInstance("RSA").generatePublic(X509EncodedKeySpec(x509key));
var cip = Cipher.getInstance("RSA/ECB/PKCS1Padding");
cip.init(Cipher.ENCRYPT_MODE, pubKey);
bjson = json.getBytes("UTF-8");
bjson = cip.doFinal(bjson);
if (sjson = Base64.encodeToString(bjson, 0)) {
	sjson = "";
}

// ----==----
var data = sjson

// signing data with SHA256WithRSA or RSA_SHA256
// result base64 encoded
var privateKeyString = "MIIEowIBAAKCAQEA6spowYp8cYY7eEzbxdqJzAVO8sIPYZFgCEJMJtCuNFgqRHGiYVFnAEHq7VUYWGWBwy6680PQ/8jhxVZLU8uemhGQdRd378HyIbyqYWbipioiBo/izRB1FIix+z73ltMS6jsYeKskHLIXhAfpzUMg58Bfz9z4K6uTgG0NcV097xYQ2LBL9Jgz7ZHFRrL4tm3HEwSAyoYg3D+zvmPlepuDzpmdneA2IOaPHigCQu9oU3tQFJV0suC7ngsKMgm0x7KshrzP4baIpDLl644ol8YhJApa1jLO4ZLeoFYmOwAOxb2r5Gv9JIaco8du1sxVCBMItEbHeAOp12Mi4LrtnzDg/QIDAQABAoIBAQCM94H10ePtdR53mA3LuO5K5FUGQ6VgexwhWUzHcTtmihi2h1NrAqjI6srWO+UmDavfOEteizUMnduP08PeR76t2iXVZ4XHGbb6sn/JgzBJe7p6HiGZ5WumLHxP0PpuP8R3hYunx3kjEXo091tUth0nh8T+1DjcX2NsFA2iJs1sobhsEdj9FJqk58tw5YLAQMhpGN5/fvpnpxJ++zKxUiZQuxaYB2JBURblt1EcXl1qjaYRyDkPT3JjTrm74dZkFcI67fbe/j+TbZ08/8V6wmZkXkLrtyrT3iHdrQ/vckxiA05p92jiGSiqYZjvvSYYPiIplvw89R6QbbVDI1WerNOBAoGBAPyjiV7ZtFBj8DJTGLVl7xTsysw/lm6uUiloh4OB+NOyQRp57NEY/NsHx0oj7RJMcHJVPbB6yHzadpaN0tYlocNpEJvDoJFi3lDdK8IQEDeR26c02brHewlagHQR+YSZCQAH6fhkmZtWeiWT06C+IYzlzYacZbVA6QDi3TzKcyiFAoGBAO3qFWEGFiSxQAh46xis4kKY7oYjov2HQmQrqiUmOXLOFPU10QchjQ5SMWo07MbCIu89SPdcFTce+DjB/SIZ7rnMqHhsv+nDhY4x3/N8mVs6CMNWBTZHKSdoCZqNkYk8WSrgjkiPF1QWgZloTXqph1df3QSqDuCkqJMw/+hag/wZAoGAQm19oUL89mZQserHhGrZAleAu1O86j1LvTeO6jUyh/dGxv9oFVLMqYrSC9yGaVlP98ASCGVAchO/OrbvyP78GbV2eH5nqdKlQf3lk5FFslbXuo9a8JGi+EKjo+FWCldy1l8+VeGdwgpq85Z6/rj/eVyzPcHBvHbkzSbZpTJWNUECgYApk7fqvcKO5P74vBDxvfzIgKtme8HXJpUeplbUqSpu+r7E34PJCD77WrfSBK/ASFsMjhw7Rjda66mbWaNgG0IR7EcgzHh5ZLbpU6P8jDDMtYakhFri0mrJvS+4d7cQUaCPkrCq037TLh+17Hllz0yco18fmoyBC6V+G5kNRQw9uQKBgDOhVbFPYVIwpuvFxytZh+aATYrLCkVw3IHmuuZMqaZSYHip2Tn1nyokpYnujMljJXCnXIFivejCMvX35HUpTU8NJ9laoA2Vixw1t+qtYVjYFXUnJoOlEcX29KuTNPkPNm3GwQd9cQ1XphuL8yrp0pxnvOfBYixOX5sFCjfMBrXN"
var privKey = KeyFactory.getInstance("RSA").generatePrivate(PKCS8EncodedKeySpec(privateKeyString));
var sig = Signature.getInstance("SHA256WithRSA");
sig.initSign(privKey);
var bdata = data.getBytes("UTF-8");
sig.update(bdata);
var bsig = sig.sign();
Base64.encodeToString(bsig, 10)

```

**Reply by derodevil** [reply]
gagal di sini. muncul pesan `Bad Request: Invalid decrypted string`. value kyc itu apa ya gan? sama signFirstTime apakah ini boolean?

**Reply by lintangtimur** [reply]
signFirstTime tipe long
kyc belum tau

**Reply by derodevil** [reply]
Sy cek apk value kyc itu "Y" atau "N". Sy coba json (unformatted) seperti berikut

```
{
	"deviceId": "047d3ddf-2dc4-413b-b742-f9eabdf67717",
	"kyc: "Y",
	"phoneNumber": "08123456789",
	"signFirstTime": 1624690546152
}
```
Json di-rsa sha256 lalu hasilnya signed dan responnya masih tetep `Bad Request: Invalid decrypted string`

**Reply by lintangtimur** [reply]
java.security.InvalidKeyException: IOException : algid parse error, not a sequence

```java
KeyFactory instance = KeyFactory.getInstance("RSA");

            String private_key_string = "MIIEowIBAAKCAQEA6spowYp8cYY7eEzbxdqJzAVO8sIPYZFgCEJMJtCuNFgqRHGiYVFnAEHq7VUYWGWBwy6680PQ/8jhxVZLU8uemhGQdRd378HyIbyqYWbipioiBo/izRB1FIix+z73ltMS6jsYeKskHLIXhAfpzUMg58Bfz9z4K6uTgG0NcV097xYQ2LBL9Jgz7ZHFRrL4tm3HEwSAyoYg3D+zvmPlepuDzpmdneA2IOaPHigCQu9oU3tQFJV0suC7ngsKMgm0x7KshrzP4baIpDLl644ol8YhJApa1jLO4ZLeoFYmOwAOxb2r5Gv9JIaco8du1sxVCBMItEbHeAOp12Mi4LrtnzDg/QIDAQABAoIBAQCM94H10ePtdR53mA3LuO5K5FUGQ6VgexwhWUzHcTtmihi2h1NrAqjI6srWO+UmDavfOEteizUMnduP08PeR76t2iXVZ4XHGbb6sn/JgzBJe7p6HiGZ5WumLHxP0PpuP8R3hYunx3kjEXo091tUth0nh8T+1DjcX2NsFA2iJs1sobhsEdj9FJqk58tw5YLAQMhpGN5/fvpnpxJ++zKxUiZQuxaYB2JBURblt1EcXl1qjaYRyDkPT3JjTrm74dZkFcI67fbe/j+TbZ08/8V6wmZkXkLrtyrT3iHdrQ/vckxiA05p92jiGSiqYZjvvSYYPiIplvw89R6QbbVDI1WerNOBAoGBAPyjiV7ZtFBj8DJTGLVl7xTsysw/lm6uUiloh4OB+NOyQRp57NEY/NsHx0oj7RJMcHJVPbB6yHzadpaN0tYlocNpEJvDoJFi3lDdK8IQEDeR26c02brHewlagHQR+YSZCQAH6fhkmZtWeiWT06C+IYzlzYacZbVA6QDi3TzKcyiFAoGBAO3qFWEGFiSxQAh46xis4kKY7oYjov2HQmQrqiUmOXLOFPU10QchjQ5SMWo07MbCIu89SPdcFTce+DjB/SIZ7rnMqHhsv+nDhY4x3/N8mVs6CMNWBTZHKSdoCZqNkYk8WSrgjkiPF1QWgZloTXqph1df3QSqDuCkqJMw/+hag/wZAoGAQm19oUL89mZQserHhGrZAleAu1O86j1LvTeO6jUyh/dGxv9oFVLMqYrSC9yGaVlP98ASCGVAchO/OrbvyP78GbV2eH5nqdKlQf3lk5FFslbXuo9a8JGi+EKjo+FWCldy1l8+VeGdwgpq85Z6/rj/eVyzPcHBvHbkzSbZpTJWNUECgYApk7fqvcKO5P74vBDxvfzIgKtme8HXJpUeplbUqSpu+r7E34PJCD77WrfSBK/ASFsMjhw7Rjda66mbWaNgG0IR7EcgzHh5ZLbpU6P8jDDMtYakhFri0mrJvS+4d7cQUaCPkrCq037TLh+17Hllz0yco18fmoyBC6V+G5kNRQw9uQKBgDOhVbFPYVIwpuvFxytZh+aATYrLCkVw3IHmuuZMqaZSYHip2Tn1nyokpYnujMljJXCnXIFivejCMvX35HUpTU8NJ9laoA2Vixw1t+qtYVjYFXUnJoOlEcX29KuTNPkPNm3GwQd9cQ1XphuL8yrp0pxnvOfBYixOX5sFCjfMBrXN";
            System.out.println(private_key_string);
            byte[] decode = Base64.decode(private_key_string, 0);

            PrivateKey generatePrivate = instance.generatePrivate(new PKCS8EncodedKeySpec(decode));

            Signature instance2 = Signature.getInstance("SHA256WithRSA");
            instance2.initSign(generatePrivate);
            byte[] bytes = string.getBytes(StandardCharsets.UTF_8);
            instance2.update(bytes);
            byte[] sign = instance2.sign();

            strSign = Base64.encodeToString(sign, 10);
```
pas sign malah error, tau dimana errornya?? penulisan privateKey apakah benar?

**Reply by derodevil** [reply]
Saya ga pake java tp pake C#. Kurang tau dg error di java

**Reply by namtxs** [reply]
{
	"security_code": "<security_code>",
}

**Reply by derodevil** [reply]
security_code ini maksdunya ditambahkan ke DataSharing yg json itu kah?

**Reply by SilentX26** [reply]
ada yang udah bisa ngatasin error **Bad Request: Invalid decrypted string**?

**Comment by ii64** [comment]
Key awal pake apa ini?
![image](https://user-images.githubusercontent.com/26342994/125522015-6bd7d2e6-0e8c-42dd-a0c6-b08659867711.png)

**Reply by lintangtimur** [reply]
sepertinya ga ada isinya

**Comment by SilentX26** [comment]
signFirstTime tipe data nya long, value nya ini time disaat pertama kita request otp atau apa ya?

**Reply by lintangtimur** [reply]
pada waktu HIT api ini mungkin

**Comment by SilentX26** [comment]
### Flow enkripsi accounts/login
Kemarin saya sudah baca-baca decompile aplikasi ovo, saya menemukan flow berikut (sudah di test dan hasilnya success):

1. data yang di enkripsi berupa string dengan pola berikut: _**"LOGIN|" + securityCode + "|" + currentTimeMillies + "|" + deviceId + "|" + phoneNumber + "|" + deviceId + "|" + otpRefId**_. 
2.  Keterangan: 
    - **securityCode**: berupa PIN akun ovo anda
    - **currentTimeMillies**: berupa waktu (dalam milisecond) pada saat anda hit ke API OVO
    - **deviceId**: berupa device id anda
    - **phoneNumber**: berupa nomor HP anda
    - **otpRefId**: berupa ref id otp yang didapat pada saat request OTP
2. Contoh: **"LOGIN|123456|1628222553987|abcdefgh|+628xxxxxxx|abcdefgh|xxxxxxxxxxx"**
3. data di enkripsi menggunakan public key yang didapat dari **user/public_keys**
3. di encoding menggunakan base64


![Screenshot (86)](https://user-images.githubusercontent.com/55566609/128453827-1cbfba6e-d281-42cf-86c1-bf2ca8665c29.png)

**Reply by derodevil** [reply]
> 3\. xxxxxx

Makasih mas, mantuuul :)

**Reply by SilentX26** [reply]
> Makasih mas, mantuuul :)

Sama sama gan

**Reply by lintangtimur** [reply]
```
{
	"deviceId": "047d3ddf-2dc4-413b-b742-f9eabdf67717",
	"kyc: "Y",
	"phoneNumber": "08123456789",
	"signFirstTime": 1624690546152
}
```
kalo ini apa ya? solanya nemu juga di parameter ovo bagian encryptan.

**Reply by SilentX26** [reply]
> ```
> {
> 	"deviceId": "047d3ddf-2dc4-413b-b742-f9eabdf67717",
> 	"kyc: "Y",
> 	"phoneNumber": "08123456789",
> 	"signFirstTime": 1624690546152
> }
> ```
> 
> kalo ini apa ya? solanya nemu juga di parameter ovo bagian encryptan.

kurang tau kalo itu gan, tpi sepertinya untuk berkomunikasi dengan third party.

**Reply by hiibnu** [reply]
response yang saya dapatkan 
{
    "data": null,
    "response_code": "OV00002",
    "response_message": "Bad Request: Invalid decrypted string"
}

dengan encrypt :
var buffer = Buffer.from(toEncrypt);
var encrypted = crypto.publicEncrypt(rsa, "LOGIN|123456|1628222553987|abcdefgh|+628xxxxxxx|abcdefgh|xxxxxxxxxxx");
encrypted.toString("base64");

kira-kira dimana yang salah ya? terimakasih.

**Reply by SilentX26** [reply]
> response yang saya dapatkan
> {
> "data": null,
> "response_code": "OV00002",
> "response_message": "Bad Request: Invalid decrypted string"
> }
> 
> dengan encrypt :
> var buffer = Buffer.from(toEncrypt);
> var encrypted = crypto.publicEncrypt(rsa, "LOGIN|123456|1628222553987|abcdefgh|+628xxxxxxx|abcdefgh|xxxxxxxxxxx");
> encrypted.toString("base64");
> 
> kira-kira dimana yang salah ya? terimakasih.

Coba paste base64 nya kesini gan

**Reply by hiibnu** [reply]
udah bisa gan makasih. encryptnya pake python dari @lintangtimur di issue sebelah. untuk api cek balance gitu udah beda lagi ya ?

**Reply by hiibnu** [reply]
ini base64 yang di enc pake js gan :
ukvpbPOBgHn7OxOEpvM30BZTySnf8F/I19OE+TSMQPnPYHxkwqOMK+R5nIAcTV0kZvEzl4mvOOEr4b0udaC9Z9muk5IMAB99hms57iicBiIuCdilY6fSuOCrkHrQ8jiHpGlb9JrRnfY5AQ7155Rg8Yz1Mc8cjh/Sg3DgrzNPnc4q4Qfxbs0oXIw/vQbTM0cMkke4fkGkvNTe8r76Li4N0ZeBeSUzLuONASCLOQOVBsMN05b535NiinUgnJh5dahF61m1G+ONO6K7Y1jakGAjtY/5juySXnafILviZWEIB8YbH5IBjbN+IzRZyUemg5eWCihu1izTCn8+xZbCRPcroQ==

**Reply by lintangtimur** [reply]
> udah bisa gan makasih. encryptnya pake python dari @lintangtimur di issue sebelah. untuk api cek balance gitu udah beda lagi ya ?

koreksi, solusi python dari @herihermwn.
Sudah ditest? saya belum bisa test karna belum dapat auth tokennya

**Reply by hiibnu** [reply]
Sudah saya tes, dan ini resultnya,
{
    "response_code": "OV00000",
    "response_message": "Success",
    "data": {
        "auth": {
            "access_token": "eyJhbGciOiJSUzI1NiJ9.eyJleHBpcnlJbk1pbGxpU2Vjb25kcyI6NjA0ODAwMDAwLCJjcmVhdGVUaW1lIjoxNjI5NjUwNDA3MDA2LCJzZWNyZXQiOiJod3Y3ZHFtdWMrekQ1MjBaRGVBemZsQS8zeGhKQ3N6WE91aFZ4Wk9EbjV2ZTZxTXlTbnk0b0lMTENMM0d4UnFsMjZNeXJYLzZFdDF1V0FHY3pLQ0dzb3VMM0hzU01zdjZIOUkveVFXYThDOFZTUW1JZ2FZQkx5aTFLT2NaVmYySmNaUkk3V2gxWk5nSEt0QVFiSzVoVkthMFp1NUozZzg4L2VURE1oWWZZWWY0K3RHL2VDS1Qxalg5R0pWSitOSjhyaTV1SlI0RzhvUFIvUzArOUk4MkxlMG1RYnYwTmNVYTd2T09GZ1hWTncvRVJBNGxZdEdWNTNVZnUvaFZQWU1qNVQzSnpKYVhxZittbU5MUU1TTGpZaDNNWGUyb1VnVlhmaHZ1UVlHY1hUODZQZmJTQWJtWnpHT1cwQXd1N2xLRnFJL0JrWHZVOEhmRHFXbUZtbCtZNlU5UHg3Uml6WjJTODJ2b2JGTEM4UGM9In0.iODrRf1J1IkUh-cUZW7zEU9ekSyOIAobPCO9xTkQVfcg5hf1QbwhAPdx4rQYBibKwo1luZLaaDXyuW9MmhA7yZxI6MVEKia7Fhb2CsXgis3vtopMtCNkUi1lK7FEOxBDkJyg9e93arRqkQO7s9yB_ELJJ5uU81vUeZC0fE3bjn8iO9azCQax8LnJjRTNcYTvH24o-xxxxxx-xl0PXNhT47fRVS8gsCLd4aMrX_1fAD9BXPXPP1Wq-Al_ZOlJZScN2m2Hj0LbzBei9WMRg3mY6Tt8lja6xDAxwKwoY_nITRKfLW-6oK8RwJ43aIfA40JLyn2_xDY1fhR0KEhLuMTQ",
            "token_type": "Bearer",
            "expires_in": 1629736807,
            "refresh_token": "eyJhbGciOiJSUzI1NiJ9.eyJleHBpcnlJbk1pbGxpU2Vjb25kcyI6NjA0ODAwMDAwLCJjcmVhdGVUaW1lIjoxNjI5NjUwNDA3MDA2LCJzZWNyZXQiOiJod3Y3ZHFtdWMrekQ1MjBaRGVBemZsQS8zeGhKQ3N6WE91aFZ4Wk9EbjV2ZTZxTXlTbnk0b0lMTENMM0d4UnFsMjZNeXJYLzZFdDF1V0FHY3pLQ0dzb3VMM0hzU01zdjZIOUkveVFXYThDOFZTUW1JZ2FZQkx5aTFLT2NaVmYySmNaUkk3V2gxWk5nSEt0QVFiSzVoVkthMFp1NUozZzg4L2VURE1oWWZZWWY0K3RHL2VDS1Qxalg5R0pWSitOSjhyaTV1SlI0RzhvUFIvUzArOUk4MkxlMG1RYnYwTmNVYTd2T09GZ1hWTncvRVJBNGxZdEdWNTNVZnUvaFZQWU1qNVQzSnpKYVhxZittbU5MUU1TTGpZaDNNWGUyb1VnVlhmaHZ1UVlHY1hUODZQZmJTQWJtWnpHT1cwQXd1N2xLRnFJL0JrWHZVOEhmRHFXbUZtbCtZNlU5UHg3Uml6WjJTODJ2b2JGTEM4UGM9In0.iODrRf1J1IkUh-cUZW7zEU9ekSyOIAobPCO9xTkQVfcg5hf1QbwhAPdx4rQYBibKwo1luZLaaDXyuW9MmhA7yZxI6MVEKia7Fhb2CsXgis3vtopMtCNkUi1lK7FEOxBDkJyg9e93arRqkQO7s9yB_ELJJ5uU81vUeZC0fE3bjn8iO9azCQax8LnJjRTNcYTvH24o-xxxxxx-xl0PXNhT47fRVS8gsCLd4aMrX_1fAD9BXPXPP1Wq-Al_ZOlJZScN2m2Hj0LbzBei9WMRg3mY6Tt8lja6xDAxwKwoY_nITRKfLW-6oK8RwJ43aIfA40JLyn2_xDY1fhR0KEhLuMTQ",
            "scope": "user"
        },
        "account": {
            "ovo_id": "000110xxxxxxxxx",
            "msisdn": "0822xxxxxx",
            "account_status": "REGISTERED",
            "kyc_status": "",
            "authentication": "PROVIDER_OTP_PASSWORD_ENCRYPTED",
            "push_notification_id": "fs-DYcGaRbKERLhF4hkQ92:APA91bEjjUFzzFvadIKtdrrqsyrGH26xLRR5-Oyym2l9Ybv0O1cnvqA14ghuTbXz0ogazN-Kw6iGxW2klakANBaVXoFCLrT4hWJJ5FCGOz2o5bGE7RX6XpxndNkcxnqpWat449vBvYSa"
        }
    }
}

tapi ketika saya hit endpoint https://api.ovo.id/v1.0/api/front/ , response yang sy dapatkan Not secure token detected,
{
    "code": 1025,
    "content": {
        "present": false
    },
    "message": "Not secure token detected"
}

**Reply by herihermwn** [reply]
> Sudah saya tes, dan ini resultnya,
> {
> "response_code": "OV00000",
> "response_message": "Success",
> "data": {
> "auth": {
> "access_token": "eyJhbGciOiJSUzI1NiJ9.eyJleHBpcnlJbk1pbGxpU2Vjb25kcyI6NjA0ODAwMDAwLCJjcmVhdGVUaW1lIjoxNjI5NjUwNDA3MDA2LCJzZWNyZXQiOiJod3Y3ZHFtdWMrekQ1MjBaRGVBemZsQS8zeGhKQ3N6WE91aFZ4Wk9EbjV2ZTZxTXlTbnk0b0lMTENMM0d4UnFsMjZNeXJYLzZFdDF1V0FHY3pLQ0dzb3VMM0hzU01zdjZIOUkveVFXYThDOFZTUW1JZ2FZQkx5aTFLT2NaVmYySmNaUkk3V2gxWk5nSEt0QVFiSzVoVkthMFp1NUozZzg4L2VURE1oWWZZWWY0K3RHL2VDS1Qxalg5R0pWSitOSjhyaTV1SlI0RzhvUFIvUzArOUk4MkxlMG1RYnYwTmNVYTd2T09GZ1hWTncvRVJBNGxZdEdWNTNVZnUvaFZQWU1qNVQzSnpKYVhxZittbU5MUU1TTGpZaDNNWGUyb1VnVlhmaHZ1UVlHY1hUODZQZmJTQWJtWnpHT1cwQXd1N2xLRnFJL0JrWHZVOEhmRHFXbUZtbCtZNlU5UHg3Uml6WjJTODJ2b2JGTEM4UGM9In0.iODrRf1J1IkUh-cUZW7zEU9ekSyOIAobPCO9xTkQVfcg5hf1QbwhAPdx4rQYBibKwo1luZLaaDXyuW9MmhA7yZxI6MVEKia7Fhb2CsXgis3vtopMtCNkUi1lK7FEOxBDkJyg9e93arRqkQO7s9yB_ELJJ5uU81vUeZC0fE3bjn8iO9azCQax8LnJjRTNcYTvH24o-xxxxxx-xl0PXNhT47fRVS8gsCLd4aMrX_1fAD9BXPXPP1Wq-Al_ZOlJZScN2m2Hj0LbzBei9WMRg3mY6Tt8lja6xDAxwKwoY_nITRKfLW-6oK8RwJ43aIfA40JLyn2_xDY1fhR0KEhLuMTQ",
> "token_type": "Bearer",
> "expires_in": 1629736807,
> "refresh_token": "eyJhbGciOiJSUzI1NiJ9.eyJleHBpcnlJbk1pbGxpU2Vjb25kcyI6NjA0ODAwMDAwLCJjcmVhdGVUaW1lIjoxNjI5NjUwNDA3MDA2LCJzZWNyZXQiOiJod3Y3ZHFtdWMrekQ1MjBaRGVBemZsQS8zeGhKQ3N6WE91aFZ4Wk9EbjV2ZTZxTXlTbnk0b0lMTENMM0d4UnFsMjZNeXJYLzZFdDF1V0FHY3pLQ0dzb3VMM0hzU01zdjZIOUkveVFXYThDOFZTUW1JZ2FZQkx5aTFLT2NaVmYySmNaUkk3V2gxWk5nSEt0QVFiSzVoVkthMFp1NUozZzg4L2VURE1oWWZZWWY0K3RHL2VDS1Qxalg5R0pWSitOSjhyaTV1SlI0RzhvUFIvUzArOUk4MkxlMG1RYnYwTmNVYTd2T09GZ1hWTncvRVJBNGxZdEdWNTNVZnUvaFZQWU1qNVQzSnpKYVhxZittbU5MUU1TTGpZaDNNWGUyb1VnVlhmaHZ1UVlHY1hUODZQZmJTQWJtWnpHT1cwQXd1N2xLRnFJL0JrWHZVOEhmRHFXbUZtbCtZNlU5UHg3Uml6WjJTODJ2b2JGTEM4UGM9In0.iODrRf1J1IkUh-cUZW7zEU9ekSyOIAobPCO9xTkQVfcg5hf1QbwhAPdx4rQYBibKwo1luZLaaDXyuW9MmhA7yZxI6MVEKia7Fhb2CsXgis3vtopMtCNkUi1lK7FEOxBDkJyg9e93arRqkQO7s9yB_ELJJ5uU81vUeZC0fE3bjn8iO9azCQax8LnJjRTNcYTvH24o-xxxxxx-xl0PXNhT47fRVS8gsCLd4aMrX_1fAD9BXPXPP1Wq-Al_ZOlJZScN2m2Hj0LbzBei9WMRg3mY6Tt8lja6xDAxwKwoY_nITRKfLW-6oK8RwJ43aIfA40JLyn2_xDY1fhR0KEhLuMTQ",
> "scope": "user"
> },
> "account": {
> "ovo_id": "000110xxxxxxxxx",
> "msisdn": "0822xxxxxx",
> "account_status": "REGISTERED",
> "kyc_status": "",
> "authentication": "PROVIDER_OTP_PASSWORD_ENCRYPTED",
> "push_notification_id": "fs-DYcGaRbKERLhF4hkQ92:APA91bEjjUFzzFvadIKtdrrqsyrGH26xLRR5-Oyym2l9Ybv0O1cnvqA14ghuTbXz0ogazN-Kw6iGxW2klakANBaVXoFCLrT4hWJJ5FCGOz2o5bGE7RX6XpxndNkcxnqpWat449vBvYSa"
> }
> }
> }
> 
> tapi ketika saya hit endpoint https://api.ovo.id/v1.0/api/front/ , response yang sy dapatkan Not secure token detected,
> {
> "code": 1025,
> "content": {
> "present": false
> },
> "message": "Not secure token detected"
> }

Coba ganti url APInya jadi https://api.ovo.id/v3.0/api/front/. 

dan pastikan headernya kaya gini : 
{
        'client-id': 'ovo_android',
        'device-id': device_id,
        'authorization': refresh_token,
        'Connection': 'close',
        'Accept-Encoding': 'gzip, deflate',
        'User-Agent': 'okhttp/4.9.0',
        'Content-Type': 'application/json'
}

**Reply by hiibnu** [reply]
makasih mas, endpointnya bisa

**Reply by apriady** [reply]
mas @notme1001 lagi bikin client buat nodejs kah itu?

**Reply by hiibnu** [reply]
iya mas, lagi nyoba" sih ini @apriady .

**Reply by apriady** [reply]
boleh share bagian enkripsi menggunakan publickeynya mas @notme1001 ?
ini kebetulan saya kemarin masih stuck di sana.

**Reply by hiibnu** [reply]
Encryptnya saya pake versi pythonnya dari mas @herihermwn , untuk jsnya sy blm cb lg

**Reply by hiibnu** [reply]
@apriady untuk encrypt jsnya sy udh coba pake ini , berhasil

![carbon](https://user-images.githubusercontent.com/57023894/131830347-aa7c1117-140f-4ce2-8453-d911c3bd9725.png)

**Reply by ipkzone** [reply]
mas mau tanya kalo untuk registrasinya gmna ya untuk pola encript nya
apa aja ya yang di encript untuk registrasi akun nya
